### PR TITLE
BlockOperator and BlockDiagonalOperator

### PR DESCRIPTION
**Pull request by pmli**
Code: src/pymor/playground/block.py
Tests: src/pymortests/block.py

### CODE REVIEW

File: src/pymor/playground/operators/block.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+                range_types[i] = op.range
+
+        self.source = VectorSpace(BlockVectorArray, tuple(source_types))
+        self.range = VectorSpace(BlockVectorArray, tuple(range_types))
+        self._source_dims = tuple(space.dim for space in self.source.subtype)
+        self._range_dims = tuple(space.dim for space in self.range.subtype)
+        self.num_source_blocks = len(source_types)
+        self.num_range_blocks = len(range_types)
+        self.linear = all(op.linear for op in self._operators())
+        self._is_diagonal = (all(block is None if i != j else True for (i, j), block in np.ndenumerate(self._blocks))
```
Comment (ftalbrecht) [inline src/pymor/playground/operators/block.py]: `is_diagonal` is not used

File: src/pymor/playground/operators/block.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+            block = None
+            for i in xrange(self.num_range_blocks):
+                op = self._blocks[i, j]
+                if op is not None:
+                    V = op.apply_adjoint(U.block(i), ind=ind, mu=mu)
+                    if block is None:
+                        block = V.copy()
+                    else:
+                        block += V
+            blocks.append(block.copy())
```
Comment (ftalbrecht) [inline src/pymor/playground/operators/block.py]: s.a.

File: src/pymor/playground/operators/block.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        n = len(blocks)
+        blocks2 = np.array([[None for j in xrange(n)] for i in xrange(n)])
+        for i, op in enumerate(blocks):
+            blocks2[i, i] = op
+        super(BlockDiagonalOperator, self).__init__(blocks2)
+
+    def apply_inverse(self, V, ind=None, mu=None, least_squares=False):
+        U = []
+        for i in xrange(self.num_source_blocks):
+            U.append(self._blocks[i, i].apply_inverse(V.block(i), ind=ind, mu=mu, least_squares=least_squares))
```
Comment (ftalbrecht) [inline src/pymor/playground/operators/block.py]: s.a.

**Comment by ftalbrecht** [comment]
Sorry for being so late. I left a few notes, please take a look. Apart from that: looks good to me!

File: src/pymor/playground/operators/block.py (hunk 1 above)
Comment (pmli) [inline src/pymor/playground/operators/block.py]: Agreed. Should it be used somehow or can it be removed?

File: src/pymor/playground/operators/block.py (hunk 2 above)
Comment (pmli) [inline src/pymor/playground/operators/block.py]: s.a. = ? :confused:

File: src/pymor/playground/operators/block.py (hunk 1 above)
Comment (ftalbrecht) [inline src/pymor/playground/operators/block.py]: I guess I would just remove it...

File: src/pymor/playground/operators/block.py (hunk 2 above)
Comment (ftalbrecht) [inline src/pymor/playground/operators/block.py]: Ah, sorry for that. I meant s.a. as in see above. As in: 
* if we do not need `copy()` above, then we do not need it here, either
* if you agree that a pre-initialization would make sense above, then I would also pre-initialize `blocks` as a list of appropriate length (`num_source_blocks`) here. The same holds further below...

**Comment by sdrave** [comment]
Looks good to me! When merging I will only update the copyright notice and remove the import from the `__init__.py` (we have the policy of not importing subpackages). I believe we could also directly move this from the playground to `pymor.operators.block`?

**Comment by ftalbrecht** [comment]
Yes, no need to keep this in playground.

**Comment by sdrave** [comment]
Hi @pmli, thank you very much for your contribution! Your code is now in master and I have added an attribution to `AUTHORS.md`. 

By adding fixtures for BlockOperator and BlockDiagonalOperator to our testing framework, I discovered a small bug in your `BlockDiagonalOperator.apply_inverse_adjoint` implementation in the case when scalar products are present: The inverse adjoint operator maps from source to range, so the `source_product` needs to be applied first, not the `range_product`. I simplified the code by calling the default implementation which takes care of the scalar products, so only the no-product case has to be implemented.

**Pull request merged**